src/deploy/functions/prepare.ts

```diff
@@ -51,10 +80,21 @@ export async function prepare(
   const firebaseConfig = await functionsConfig.getFirebaseConfig(options);
   context.firebaseConfig = firebaseConfig;
   const runtimeConfig = await getFunctionsConfig(context);
-  const env = await getEnvs(context);
+
+  utils.assertDefined(
+    options.config.src.functions.source,
```

Comment:
Why did you add this assertion? Is this whole thing safe to do without optional chaining? What happens if you don't have functions set up but you're deploying e.g. hosting, won't this throw?

Reply:
This was meant to move existing assertion made couple of lines below this one (while rebasing, looks like I forgot to delete the existing assertion).

https://github.com/firebase/firebase-tools/blob/98ff7f72f9abad53dc2d675b61171f59382df90f/src/deploy/functions/prepare.ts#L79-L83

IIUC, behavior today is that if your firebase.json does not have `functions` attribute, it won't attempt to deploy any functions hence will never hit this code. Is there a different scenario you are referring to here?

Reply:
On second scan, this check seems unnecessary since we always provide a default for `functions.source` when missing:

https://github.com/firebase/firebase-tools/blob/98ff7f72f9abad53dc2d675b61171f59382df90f/src/config.ts#L78-L86

I'm going to opt on removing this check all togther.